server/src/internalClusterTest/java/org/elasticsearch/snapshots/SnapshotThrottlingIT.java

```diff
@@ -0,0 +1,173 @@
+import org.elasticsearch.test.ESIntegTestCase;
+import org.elasticsearch.test.MockLogAppender;
+import org.elasticsearch.test.junit.annotations.TestLogging;
+
+import java.util.Collections;
+
+import static org.elasticsearch.indices.recovery.RecoverySettings.INDICES_RECOVERY_MAX_BYTES_PER_SEC_SETTING;
+import static org.elasticsearch.indices.recovery.RecoverySettings.NODE_BANDWIDTH_RECOVERY_DISK_READ_SETTING;
+import static org.elasticsearch.indices.recovery.RecoverySettings.NODE_BANDWIDTH_RECOVERY_DISK_WRITE_SETTING;
+import static org.elasticsearch.indices.recovery.RecoverySettings.NODE_BANDWIDTH_RECOVERY_NETWORK_SETTING;
+import static org.hamcrest.Matchers.greaterThan;
+
+@ESIntegTestCase.ClusterScope(numDataNodes = 0, scope = ESIntegTestCase.Scope.TEST)
+public class SnapshotThrottlingIT extends AbstractSnapshotIntegTestCase {
+
+    private Tuple<Long, Long> testThrottledRepository(String throttleSnapshot, String throttleRestore, boolean compressRepo)
+        throws Exception {
+        logger.info(
+            "--> testing throttled repository (throttleSnapshot=[{}], throttleRestore=[{}], compressRepo=[{}])",
+            throttleSnapshot,
+            throttleRestore,
+            compressRepo
+        );
+        createRepository(
+            "test-repo",
+            "fs",
+            Settings.builder()
+                .put("location", randomRepoPath())
+                .put("compress", compressRepo)
+                .put("chunk_size", randomIntBetween(1000, 4000), ByteSizeUnit.BYTES)
+                .put("max_snapshot_bytes_per_sec", throttleSnapshot)
+                .put("max_restore_bytes_per_sec", throttleRestore)
+        );
+        createSnapshot("test-repo", "test-snap", Collections.singletonList("test-idx"));
+        RestoreSnapshotResponse restoreSnapshotResponse = client().admin()
+            .cluster()
+            .prepareRestoreSnapshot("test-repo", "test-snap")
+            .setRenamePattern("test-")
+            .setRenameReplacement("test2-")
+            .setWaitForCompletion(true)
+            .execute()
+            .actionGet();
+        assertThat(restoreSnapshotResponse.getRestoreInfo().totalShards(), greaterThan(0));
+        assertDocCount("test-idx", 100L);
+        long snapshotPause = 0L;
+        long restorePause = 0L;
+        for (RepositoriesService repositoriesService : internalCluster().getDataNodeInstances(RepositoriesService.class)) {
+            snapshotPause += repositoriesService.repository("test-repo").getSnapshotThrottleTimeInNanos();
+            restorePause += repositoriesService.repository("test-repo").getRestoreThrottleTimeInNanos();
+        }
+        cluster().wipeIndices("test2-idx");
+        logger.warn("--> tested throttled repository with snapshot pause [{}] and restore pause [{}]", snapshotPause, restorePause);
+        return new Tuple<>(snapshotPause, restorePause);
+    }
+
+    public void testThrottling() throws Exception {
```

Comment:
What do you think of having just a test for the snapshot? This change does not impact the restore rate limiter

Reply:
Thanks for the comment @tlrx ! I took another looks at the times. Indeed, if all random booleans in the test are set, then it takes 30 secs. It's because I've basically added a new experiment that asserts that the snapshots/restores happen in a lower speed of the original test. This ensures that indeed we see the relative speed drop, while the previous version only tested that pauses>0.

I've been playing a lot with the speeds here. Possibly the "10k" speed of the previous experiment was found through some experimentation and did not want to increase it too far. I've set it to 25k speed. And for the new experiment that has lower speed, I set it to 8k. Now it takes around 15 secs in worst case, and much faster on average. I've run a 100 iterations and it succeeds. We can refine it in another PR if needed.

Reply:
By the way, I run this more iterations and at one point it failed. I decreased 8k to 5k to avoid that. It succeeds now after 1000 iterations.